### PR TITLE
Added a default PortForwardManager

### PR DESCRIPTION
**Pull request by Graeme-Miller**

### CODE REVIEW

File: core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java [hunk 1]
```diff
@@ -59,46 +60,52 @@
 
     public static HostAndPort getBrooklynAccessibleAddress(Entity entity, int port) {
         String host;
-        
+
+        EntityInternal entityInternal = (EntityInternal) entity;
+
         // look up port forwarding
         PortForwardManager pfw = entity.getConfig(PORT_FORWARDING_MANAGER);
```
Comment (grkvlt) [inline core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java]: Not you, but why not change this to `pfm` instead?

File: core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java [hunk 2]
```diff
@@ -59,46 +60,52 @@
...
-            Collection<Location> ll = entity.getLocations();
-            
-            synchronized (BrooklynAccessUtils.class) {
-                // TODO finer-grained synchronization
-                
-                for (MachineLocation machine : Iterables.filter(ll, MachineLocation.class)) {
-                    HostAndPort hp = pfw.lookup(machine, port);
+        if (pfw == null) {
+            log.debug("No PortForwardManager, using default");
+            pfw = (PortForwardManager) entityInternal.getManagementContext().getLocationRegistry().getLocationManaged("portForwardManager(scope=global)");
```
Comment (grkvlt) [inline core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java]: Make `"portForwardManager(scope=global)"` a constant in the `PortForwardManager` interface.

File: core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java [hunk 3]
```diff
@@ -59,46 +60,52 @@
...
+                if (hp!=null) {
+                    log.debug("BrooklynAccessUtils found port-forwarded address {} for entity {}, port {}, using machine {}",
+                            new Object[] {hp, entity, port, machine});
+                    return hp;
+                }
+            }
+
+            Maybe<SupportsPortForwarding> supportPortForwardingLoc = Machines.findUniqueElement(ll, SupportsPortForwarding.class);
+            if (supportPortForwardingLoc.isPresent()) {
+                Cidr source = entity.getConfig(MANAGEMENT_ACCESS_CIDR);
```
Comment (grkvlt) [inline core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java]: Also `config().get(...)`

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 4]
```diff
@@ -820,32 +820,26 @@ protected MachineLocation obtainOnce(ConfigBag setup) throws NoMachinesAvailable
                 machineLocation = registerJcloudsSshMachineLocation(computeService, node, Optional.fromNullable(template), userCredentials, sshHostAndPortOverride, setup);
             }
 
+            PortForwardManager portForwardManager = setup.get(PORT_FORWARDING_MANAGER);
+            if (portForwardManager == null) {
+                LOG.debug("No PortForwardManager, using default");
+                portForwardManager = (PortForwardManager) getManagementContext().getLocationRegistry().getLocationManaged("portForwardManager(scope=global)");
```
Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: Use constant declared earlier

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 5]
```diff
@@ -2582,10 +2576,15 @@ protected void releasePortForwarding(final JcloudsMachineLocation machine) {
...
         boolean usePortForwarding = Boolean.TRUE.equals(machine.getConfig(USE_PORT_FORWARDING));
         final JcloudsPortForwarderExtension portForwarder = machine.getConfig(PORT_FORWARDER);
-        PortForwardManager portForwardManager = machine.getConfig(PORT_FORWARDING_MANAGER);
         final String nodeId = machine.getJcloudsId();
         final Map<String, Runnable> subtasks = Maps.newLinkedHashMap();
 
+        PortForwardManager portForwardManager = machine.getConfig(PORT_FORWARDING_MANAGER);
+        if (portForwardManager == null) {
+            LOG.debug("No PortForwardManager, using default");
+            portForwardManager = (PortForwardManager) getManagementContext().getLocationRegistry().getLocationManaged("portForwardManager(scope=global)");
```
Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: Again, use constant.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 6]
```diff
@@ -2617,15 +2616,10 @@ public void run() {
...
-            Set<PortMapping> mappings;
-            if (portForwardManager != null) {
-                mappings = Sets.newLinkedHashSet();
-                mappings.addAll(portForwardManager.getLocationPublicIpIds(machine));
-                if (nodeId != null) {
-                    mappings.addAll(portForwardManager.getPortMappingWithPublicIpId(nodeId));
-                }
-            } else {
-                mappings = ImmutableSet.of();
+            Set<PortMapping> mappings = Sets.newLinkedHashSet();
```
Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: Maybe use `ImmutableMap.copyOf()` here?

**Comment by grkvlt** [comment]
@duncangrant @Graeme-Miller how does this interact with #171

File: core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java [hunk 7]
```diff
@@ -59,46 +60,52 @@
...
-                for (MachineLocation machine : Iterables.filter(ll, MachineLocation.class)) {
-                    HostAndPort hp = pfw.lookup(machine, port);
+        if (pfw == null) {
+            log.debug("No PortForwardManager, using default");
+            pfw = (PortForwardManager) entityInternal.getManagementContext().getLocationRegistry().getLocationManaged("portForwardManager(scope=global)");
+        }
+
+        Collection<Location> ll = entity.getLocations();
+
+        synchronized (BrooklynAccessUtils.class) {
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java]: Do we need to check for null `pfw` here? Is it always there? What about non-jclouds locations?

Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/location/access/BrooklynAccessUtils.java]: Looking at the sources shouldn't be ever null, ignore.

**Comment by neykov** [comment]
LGTM

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java (hunk 6 above)
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: @grkvlt I don't follow - it's a `Set<PortMapping>`. What would you copy?

Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]:
```Java
Set<PortMapping> mappings = MutableMap.copyOf(portForwardManager.getLocationPublicIpIds(machine));
if (nodeId != null) {
    mappings.addAll(portForwardManager.getPortMappingWithPublicIpId(nodeId));
}
```
Yeah, it's pointless. Forget it...

**Comment by aledsage** [comment]
LGTM.
Build failure is an unrelated test, which we should mark as "Broken" in a separate PR.
Happy for Andrew's rename comments to be addressed in a separate PR in the future.

**Pull request merged**